File: src/main/services/mod-manager/path-utils.test.ts
import assert from "node:assert/strict";

import { describe, it } from "vitest";

import {
    DISABLED_PREFIX_REGEX,
    manualSubGroupSegmentMatches,
    normalizeRelativePath,
    restoreDisabledPrefix,
    stripDisabledPrefix,
} from "./path-utils.ts";

describe("DISABLED_PREFIX_REGEX", () => {
    it("matches space separator", () => {
        assert.equal(DISABLED_PREFIX_REGEX.test("DISABLED Foo"), true);
    });

    it("matches underscore separator", () => {
        assert.equal(DISABLED_PREFIX_REGEX.test("DISABLED_Foo"), true);
    });

    it("matches case-insensitively", () => {
        assert.equal(DISABLED_PREFIX_REGEX.test("disabled Foo"), true);
        assert.equal(DISABLED_PREFIX_REGEX.test("Disabled_Foo"), true);
    });

    it("does not match without separator", () => {
        assert.equal(DISABLED_PREFIX_REGEX.test("DisableFoo"), false);
        assert.equal(DISABLED_PREFIX_REGEX.test("Disable_Foo"), false);
    });

    it("does not match empty string", () => {
        assert.equal(DISABLED_PREFIX_REGEX.test(""), false);
    });

    it("does not match repeated 'disabled' without a trailing separator", () => {
        assert.equal(DISABLED_PREFIX_REGEX.test("disableddisableddisabledFoo"), false);
    });

    it("matches repeated prefixes", () => {
        assert.equal(DISABLED_PREFIX_REGEX.test("disableddisabled Foo"), true);
        assert.equal(DISABLED_PREFIX_REGEX.test("disableddisableddisabled Foo"), true);
        assert.equal(DISABLED_PREFIX_REGEX.test("disabled_disabled_Foo"), true);
        assert.equal(DISABLED_PREFIX_REGEX.test("disabled_disabled_disabled_foo"), true);
    });
});

describe("stripDisabledPrefix", () => {
    it("strips space prefix", () => {
        assert.equal(stripDisabledPrefix("DISABLED Foo"), "Foo");
    });

    it("strips underscore prefix", () => {
        assert.equal(stripDisabledPrefix("DISABLED_Foo"), "Foo");
        assert.equal(stripDisabledPrefix("disabled_My_Mod"), "My_Mod");
    });

    it("strips case-insensitively", () => {
        assert.equal(stripDisabledPrefix("disabled Foo"), "Foo");
        assert.equal(stripDisabledPrefix("Disabled_Foo"), "Foo");
    });

    it("strips trailing whitespace after name", () => {
        assert.equal(stripDisabledPrefix("DISABLED Foo  "), "Foo");
    });

    it("strips extra whitespace between prefix and name", () => {
        assert.equal(stripDisabledPrefix("DISABLED  Foo"), "Foo");
    });

    it("returns name as-is when no prefix", () => {
        assert.equal(stripDisabledPrefix("My Mod"), "My Mod");
        assert.equal(stripDisabledPrefix("DisableFoo"), "DisableFoo");
    });

    it("handles multibyte names without prefix", () => {
        assert.equal(stripDisabledPrefix("仪玄-黑珍珠"), "仪玄-黑珍珠");
    });

    it("handles multibyte names with prefix", () => {
        assert.equal(stripDisabledPrefix("DISABLED 仪玄"), "仪玄");
    });

    it("does not strip repeated 'disabled' without separator", () => {
        assert.equal(
            stripDisabledPrefix("disableddisableddisabledFoo"),
            "disableddisableddisabledFoo",
        );
    });

    it("strips repeated prefixes (double)", () => {
        assert.equal(stripDisabledPrefix("disableddisabled Foo"), "Foo");
        assert.equal(stripDisabledPrefix("disabled_disabled_Foo"), "Foo");
        assert.equal(stripDisabledPrefix("disabled disabled Foo"), "Foo");
    });

    it("strips repeated prefixes (triple)", () => {
        assert.equal(stripDisabledPrefix("disableddisableddisabled Foo"), "Foo");
        assert.equal(stripDisabledPrefix("disabled_disabled_disabled_foo"), "foo");
        assert.equal(stripDisabledPrefix("disabled disabled disabled foo"), "foo");
    });
});

describe("restoreDisabledPrefix", () => {
    it("preserves space separator from source", () => {
        assert.equal(restoreDisabledPrefix("DISABLED OldName", "NewName"), "DISABLED NewName");
    });

    it("preserves underscore separator from source", () => {
        assert.equal(restoreDisabledPrefix("DISABLED_OldName", "NewName"), "DISABLED_NewName");
    });

    it("preserves lowercase casing from source", () => {
        assert.equal(restoreDisabledPrefix("disabled OldName", "NewName"), "disabled NewName");
    });

    it("does not add prefix when source has none", () => {
        assert.equal(restoreDisabledPrefix("OldName", "NewName"), "NewName");
    });
});

describe("normalizeRelativePath", () => {
    it("strips prefix per segment and lowercases", () => {
        assert.equal(normalizeRelativePath("DISABLED Foo/Bar"), "foo/bar");
    });

    it("strips underscore prefix per segment", () => {
        assert.equal(normalizeRelativePath("DISABLED_Foo/Bar"), "foo/bar");
    });

    it("handles backslashes", () => {
        assert.equal(normalizeRelativePath("DISABLED Foo\\Bar"), "foo/bar");
    });

    it("handles mixed separators", () => {
        assert.equal(normalizeRelativePath("Foo\\DISABLED Bar/Baz"), "foo/bar/baz");
    });

    it("produces same key for enabled and disabled variants", () => {
        assert.equal(normalizeRelativePath("DISABLED MyMod"), normalizeRelativePath("MyMod"));
    });
});

describe("manualSubGroupSegmentMatches", () => {
    it("matches exact entry name", () => {
        assert.equal(manualSubGroupSegmentMatches("Foo", "foo"), true);
    });

    it("matches by stripping disabled prefix", () => {
        assert.equal(manualSubGroupSegmentMatches("DISABLED Foo", "foo"), true);
        assert.equal(manualSubGroupSegmentMatches("DISABLED_Foo", "foo"), true);
    });

    it("does not match different names", () => {
        assert.equal(manualSubGroupSegmentMatches("DISABLED Foo", "bar"), false);
    });

    it("does not match when stored segment has prefix", () => {
        assert.equal(manualSubGroupSegmentMatches("Foo", "disabled foo"), false);
    });
});

File: internal/mod/path_utils_test.go
package mod

import "testing"

func TestDisabledPrefixRegex(t *testing.T) {
	t.Parallel()

	matches := []string{
		"DISABLED Foo",
		"DISABLED_Foo",
		"disabled Foo",
		"Disabled_Foo",
		"disableddisabled Foo",
		"disableddisableddisabled Foo",
		"disabled_disabled_Foo",
		"disabled_disabled_disabled_foo",
	}
	for _, name := range matches {
		if !disabledPrefixRE.MatchString(name) {
			t.Fatalf("%q should match", name)
		}
	}

	nonMatches := []string{"DisableFoo", "Disable_Foo", "", "disableddisableddisabledFoo"}
	for _, name := range nonMatches {
		if disabledPrefixRE.MatchString(name) {
			t.Fatalf("%q should not match", name)
		}
	}
}

func TestStripDisabledPrefix(t *testing.T) {
	t.Parallel()

	cases := []struct {
		in, want string
	}{
		{"DISABLED Foo", "Foo"},
		{"DISABLED_Foo", "Foo"},
		{"disabled_My_Mod", "My_Mod"},
		{"disabled Foo", "Foo"},
		{"Disabled_Foo", "Foo"},
		{"DISABLED Foo  ", "Foo"},
		{"DISABLED  Foo", "Foo"},
		{"My Mod", "My Mod"},
		{"DisableFoo", "DisableFoo"},
		{"仪玄-黑珍珠", "仪玄-黑珍珠"},
		{"DISABLED 仪玄", "仪玄"},
		{"disableddisableddisabledFoo", "disableddisableddisabledFoo"},
		{"disableddisabled Foo", "Foo"},
		{"disabled_disabled_Foo", "Foo"},
		{"disabled disabled Foo", "Foo"},
		{"disableddisableddisabled Foo", "Foo"},
		{"disabled_disabled_disabled_foo", "foo"},
		{"disabled disabled disabled foo", "foo"},
	}
	for _, test := range cases {
		if got := stripDisabled(test.in); got != test.want {
			t.Fatalf("stripDisabled(%q) = %q, want %q", test.in, got, test.want)
		}
	}
}

func TestDisabledFileSuffix(t *testing.T) {
	t.Parallel()

	if !isDisabledFile("preview.png.disabled") {
		t.Fatal("expected preview.png.disabled to be disabled")
	}
	if !isDisabledFile("preview.png.DISABLED") {
		t.Fatal("expected preview.png.DISABLED to be disabled")
	}
	if isDisabledFile("preview.png") {
		t.Fatal("preview.png should not be disabled")
	}
	if got := stripDisabledFileSuffix("preview.png.disabled"); got != "preview.png" {
		t.Fatalf("stripDisabledFileSuffix = %q", got)
	}
	if got := stripDisabledFileSuffix("preview.png"); got != "preview.png" {
		t.Fatalf("stripDisabledFileSuffix unchanged = %q", got)
	}
}

func TestIsDisabledFolderName(t *testing.T) {
	t.Parallel()

	if !isDisabledFolderName("DISABLED Foo") || !isDisabledFolderName("DISABLED_Foo") {
		t.Fatal("expected disabled folder names")
	}
	if isDisabledFolderName("Foo") {
		t.Fatal("Foo should not be disabled")
	}
}

func TestRestoreDisabledPrefix(t *testing.T) {
	t.Parallel()

	cases := []struct {
		source, name, want string
	}{
		{"DISABLED OldName", "NewName", "DISABLED NewName"},
		{"DISABLED_OldName", "NewName", "DISABLED_NewName"},
		{"disabled OldName", "NewName", "disabled NewName"},
		{"OldName", "NewName", "NewName"},
	}
	for _, test := range cases {
		if got := restoreDisabledPrefix(test.source, test.name); got != test.want {
			t.Fatalf("restoreDisabledPrefix(%q, %q) = %q, want %q", test.source, test.name, got, test.want)
		}
	}
}

func TestNormalizeRelativePath(t *testing.T) {
	t.Parallel()

	cases := []struct {
		in, want string
	}{
		{"DISABLED Foo/Bar", "foo/bar"},
		{"DISABLED_Foo/Bar", "foo/bar"},
		{"DISABLED Foo\\Bar", "foo/bar"},
		{"Foo\\DISABLED Bar/Baz", "foo/bar/baz"},
	}
	for _, test := range cases {
		if got := normalizeRelativePath(test.in); got != test.want {
			t.Fatalf("normalizeRelativePath(%q) = %q, want %q", test.in, got, test.want)
		}
	}
	if normalizeRelativePath("DISABLED MyMod") != normalizeRelativePath("MyMod") {
		t.Fatal("enabled and disabled variants should share a key")
	}
}

func TestManualSubGroupSegmentMatches(t *testing.T) {
	t.Parallel()

	if !manualSubGroupSegmentMatches("Foo", "foo") {
		t.Fatal("expected exact match")
	}
	if !manualSubGroupSegmentMatches("DISABLED Foo", "foo") || !manualSubGroupSegmentMatches("DISABLED_Foo", "foo") {
		t.Fatal("expected stripped prefix match")
	}
	if manualSubGroupSegmentMatches("DISABLED Foo", "bar") {
		t.Fatal("different names should not match")
	}
	if manualSubGroupSegmentMatches("Foo", "disabled foo") {
		t.Fatal("stored segment with prefix should not match a bare name")
	}
}
